src/storage/types.rs

```diff
@@ -162,6 +162,30 @@ impl Key {
         let mut ts = &key[len - number::U64_SIZE..];
         number::decode_u64_desc(&mut ts)
     }
+
+    /// Whether the user key part of a ts encoded key `ts_encoded_key` equals to the encoded
+    /// user key `user_key`.
+    #[inline]
+    #[cfg_attr(feature = "cargo-clippy", allow(cast_ptr_alignment))]
+    pub fn is_user_key_eq(ts_encoded_key: &[u8], user_key: &[u8]) -> bool {
+        let user_key_len = user_key.len();
+        if ts_encoded_key.len() != user_key_len + number::U64_SIZE {
+            return false;
+        }
+        if user_key_len >= number::U64_SIZE {
```

Comment:
why check user_key last 8 bytes? seems that user_key is not a encoded key with ts.

Reply:
`user_key` is encoded but without  ts. Here it compares `user_key` with `ts_encoded_key` 's part before ts. It's correct. But there exists other problem, that is, the `unsafe` block is really unsafe..

Reply:
Consider our TiDB key encoding. For example, for table data, last 8 byte is row id and the previous bytes are same. So we compare last 8 bytes first.